### PR TITLE
Add initial support for juniper cptx

### PR DESCRIPTION
**Pull request by kmo7**
* Add initial support for juniper cptx

### CODE REVIEW

File: topo/node/cptx/cptx.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func (n *Node) DeleteNodeResource(_ context.Context, _ node.Interface) error {
+	return status.Errorf(codes.Unimplemented, "Unimplemented")
+}
+
+func defaults(pb *topopb.Node) *topopb.Node {
+	return &topopb.Node{
+		Constraints: map[string]string{
+			"cpu":    "8",
```
Comment (mhines01) [inline topo/node/cptx/cptx.go:51]: GAH that is CRAZY expensive for cpu what is up with that?

File: examples/3node-cptx-host.pb.txt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+nodes: {
+    name: "cptx"
+    type: JUNIPER_CEVO
+    config: {
+      image: "cptx:latest"
+      config_path: "/home/evo/configdisk"
+      config_file: "juniper.conf"
+      file: "../examples/cptx/cptx-config" 
+    }
+    services:{
```
Comment (mhines01) [inline examples/3node-cptx-host.pb.txt:22]: you can just take your defaults() but this doesn't hurt anything

**Review by mhines01 — COMMENTED** [review]
you need to sign the CLA as well before i can accept the PR

**Comment by kmo7** [comment]
Thanks for the review Marcus. Once the CLA is signed (more below), you can please merge the code?

Also Art is on PTO for the next week. Once he is back, would it be possible to meet with you to understand your requirements going forward?

I have been advised by our legal team that they have signed the CLA and delivered it to Google and are awaiting Google legal team's reply.
I actually tried to sign it myself, but the Google weblink rejected it as I am not an owner our mailer for this w/ Google.

Is there someone in Google our legal counsel can contact to expedite this process? They are also trying to reach out to Google in the meantime.

W.r.t. to the configuration file for our device, we will update it with a test config once our testing is done. Right now, it is just a default config as you say, and does not hurt anything.

We are also providing a CLI like other vendors whereby you can log in to our EVO CLI and do additional config at run time.

Regards,
Kaveh



Juniper Business Use Only

**Comment by kmo7** [comment]
Hi Marcus,

With  respect to your comment about why we are allocating 8 CPUs to our container, Art, who is on PTO for the next week, had communicated this a few weeks ago to Google in a meeting that other Google leads were present, but unfortunately you were not. Did they update you?

I hope you have been made aware that our solution has an EVO VM embedded in a container. Google's team had not objected to this. We are essentially providing a faithful representation of our HW platform for testing purposes. We had told Google we need 8 cores/20G of memory at the time, we are reducing that to 8 cores and 8G of memory and testing as we speak.

We can discuss this more with you when Art is back and get your inputs. I think it would be very helpful to start having discussions with you to tailor our product to your needs.

Regards,
Kaveh


Juniper Business Use Only
From: Kaveh Moezzi
Sent: Thursday, August 12, 2021 8:21 PM
To: google/kne ***@***.***>
Cc: Ganesh Raman ***@***.***>; Art Stine ***@***.***>
Subject: RE: [google/kne] Add initial support for juniper cptx (#20)

Thanks for the review Marcus. Once the CLA is signed (more below), you can please merge the code?

Also Art is on PTO for the next week. Once he is back, would it be possible to meet with you to understand your requirements going forward?

I have been advised by our legal team that they have signed the CLA and delivered it to Google and are awaiting Google legal team's reply.
I actually tried to sign it myself, but the Google weblink rejected it as I am not an owner our mailer for this w/ Google.

Is there someone in Google our legal counsel can contact to expedite this process? They are also trying to reach out to Google in the meantime.

W.r.t. to the configuration file for our device, we will update it with a test config once our testing is done. Right now, it is just a default config as you say, and does not hurt anything.

We are also providing a CLI like other vendors whereby you can log in to our EVO CLI and do additional config at run time.

Regards,
Kaveh



Juniper Business Use Only

**Comment by marcushines** [comment]
so i think all you need to do
is to follow the link in the cla page and to signup with the same email that is on your github account

**Comment by kmo7** [comment]
Hi Marcus,

Our Senior legal counsel is working with "Erin" in Google to get our CLA in place. We had to replace our existing one and that has created some additional work.
I am hoping this will get cleared by Monday.

Is the rest of the code OK for commit with the explanations I have given? We also want to follow up and meet with you please when Art gets back from PTO in a week.
There we can discuss about CPU resources as well as some questions our team has to understand your needs better.

Regards,
Kaveh



Juniper Business Use Only

**Comment by marcushines** [comment]
I just checked - Juniper does have a corporate cla with us

here are the instructions

How do I add authorized contributors for my company?
First, add their email address to the Google Group that you submitted with the corporate CLA. This is most commonly their company address.

Instruct contributors to use that same email address for any contributions to Google projects that are made on behalf of the company. For example, for projects that use git for source control, they may need to use the git config command to set their email address.

Additionally, if contributing to a project that is hosted on GitHub, contributors should add that email address to their GitHub account. This will allow us to map their GitHub username to the Corporate CLA that their contribution is covered by.


so your company should have signed up with a google group - your email needs to get added to that group and then it should be good but that will be up to whoever the administrator of your group is to get that done

**Comment by kmo7** [comment]
Ack Marcus, apparently the old CLA has to be replaced per our legal team just to get my name added as a contributor.

Ning who is our senior legal counsel is working on this. Added him to the thread.

I apologize for any inconvenience to your time this has taken. Our legal team will take care of this with your legal team.

Regards,
Kaveh



Juniper Business Use Only

**Comment by kmo7** [comment]
> 
> 
> I just checked - Juniper does have a corporate cla with us
> 
> here are the instructions
> 
> How do I add authorized contributors for my company?
> First, add their email address to the Google Group that you submitted with the corporate CLA. This is most commonly their company address.
> 
> Instruct contributors to use that same email address for any contributions to Google projects that are made on behalf of the company. For example, for projects that use git for source control, they may need to use the git config command to set their email address.
> 
> Additionally, if contributing to a project that is hosted on GitHub, contributors should add that email address to their GitHub account. This will allow us to map their GitHub username to the Corporate CLA that their contribution is covered by.
> 
> so your company should have signed up with a google group - your email needs to get added to that group and then it should be good but that will be up to whoever the administrator of your group is to get that done

**Comment by kmo7** [comment]
Hi Marcus, Google and Juniper legal cleared the CLA and added me to it. Please merge the code if you agree. Regards, Kaveh

**Comment by kmo7** [comment]
The CLA is done Marcus. Thanks, Kaveh.

**Review by marcushines — APPROVED** [review]

**Comment by marcushines** [comment]
the check. still doesn't seem to be passing - can you just create a dummy PR with just like a single line change - i just want to see if that will show the same issue?

**Comment by kmo7** [comment]
Done Marcus. Google CLA team told us we were all set on Friday. If this still does not work, please let me know and I will add you to that thread.



Juniper Business Use Only

**Review by marcushines — APPROVED** [review]

**Pull request merged**